### PR TITLE
feat: Sematext output plugin

### PR DESCRIPTION
**Pull request by bsmid**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format]

This PR adds output plugin for [Sematext](https://sematext.com). The format in which metrics are sent is similar to Influx line protocol, but Sematext platform has specific expectations from metrics data, so a fair amount of processing logic was needed to produce the right data. This logic was split into separate processors.
There are two types of processors, each represented as an interface:
- metric processors - those that process each metric in isolation
- batch processors - those that have to examine the whole batch of metrics

There are 8 processors implementing these interfaces:
- ContainerTags - injects Sematext-specific container tags (read from environment variables)
- HandleCounter - Sematext platform expects counter-type metric to be sent as delta between current and previous measurement. This processor keeps track of previous measurement and replaces original cumulative value with calculated delta.
- Heartbeat - periodically injects Sematext-specific heartbeat metric
- Host - checks Sematext-specific file on the local file-system and Sematext-specific environment variable which can override the hostname resolved by Telegraf. Sematext platform also expects a tag with name "os.host" to represent the hostname. This processor contains the logic to take care of all this.
- MetricMetainfo - Sematext platform expects metric metainfo (info about the metric name, type etc) to be sent to separate endpoint. This processor extracts the metainfo from metrics generated by Telegraf, ensures that metainfo for some metric isn't sent multiple times, and sends the metainfo batches in a separate goroutine.
- MetricType - checks for the presence of metric_type tag and removes it
- Rename - some Telegraf input integrations (nginx, mongodb, apache) are already supported by Sematext platform. There are out-of-the-box reports that expect specific names of metrics and tags. This processor renames the metrics/tags generated by Telegraf to match the expectations of Sematext platform.
- Token - injects Sematext token (specified in Telegraf config) as a tag to all metrics

A few of these transformations could have been done via config using Telegraf processors, however, that would require from user to not only enable the Sematext output plugin, but also add additional pieces of config (which may not suit other output plugins, in case the user decides to switch away from Sematext). Having all the logic encapsulated in a single output plugin seems more user-friendly and less error-prone.

I am submitting this PR on behalf of [Sematext Group](sematext.com). I signed the CLA and our CEO (Otis Gospodnetic) separately emailed the Corporate CLA.

### CODE REVIEW

**Comment by otisg** [comment]
Note that we have sent the CCLA and have received the countersigned copy, so paperwork-wise, this is cleared.

File: plugins/outputs/sematext/processors/metainfo.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+		for _, field := range metric.FieldList() {
+			mInfo, mKey := processMetric(m.token, metric, field, m.sentMetrics)
+			if mInfo != nil {
+				newMetrics[mKey] = mInfo
+			}
+		}
+	}
+
+	if len(newMetrics) > 0 {
+		go m.sendMetainfo(newMetrics)
```
Comment (srebhan) [inline plugins/outputs/sematext/processors/metainfo.go:103]: This is completely unexpected as we are in a _PROCESSOR_ and now you start sending things. I think you should collect the metadata here and return it to the caller. Then send the stuff from the main loop. This would also avoid a lot of locking, channeling, etc.

File: plugins/outputs/sematext/processors/metainfo.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+		// add newMetrics to sentMetrics so they are not sent again from other goroutines
+		for k, v := range newMetrics {
+			m.sentMetrics[k] = v
+		}
+	}
+
+	return metrics, nil
+}
+
+func (m *Metainfo) sendMetainfo(newMetrics map[string]*MetricMetainfo) {
```
Comment (srebhan) [inline plugins/outputs/sematext/processors/metainfo.go:114]: TBH this calls for a Sematext golang library. I think this code is **not** telegraf specific and others might benefit from this as well. Furthermore, just dumping the code of a client library in here has grown this PR to add >2k lines of code. Please move all _non-telegraf_ specific pieces to a library maintained in a Sematext repository to keep this plugin maintainable.

File: plugins/outputs/sematext/processors/rename.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+// Rename processor renames the measurement (metric) names
+// to match the existing metric names sent by Node.js agents
+type Rename struct{}
+
+// NewRename builds a new rename processor.
+func NewRename() BatchProcessor { return &Rename{} }
+
+// Process performs a lookup in the local maps of metric/field names
+// and replaces the metric name with the new name.
+func (r *Rename) Process(points []telegraf.Metric) ([]telegraf.Metric, error) {
```
Comment (srebhan) [inline plugins/outputs/sematext/processors/rename.go:117]: I think this processor is totally flawed. You can easily rename metrics in telegraf using the `name_override`, `name_prefix` etc functions (see [config options](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#input-plugins)). Therefore you might miss those metrics. Is there any reason why you need to "detect" the type of metric?

File: plugins/outputs/sematext/sematext.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+	defaultSematextMetricsReceiverURL = "https://spm-receiver.sematext.com"
+)
+
+// Sematext struct contains configuration read from Telegraf config and a few runtime objects.
+// We'll use one separate instance of Telegraf for each monitored service. Therefore, token for particular service
+// will be configured on Sematext output level
+type Sematext struct {
+	ReceiverURL string          `toml:"receiver_url"`
+	Token       string          `toml:"token"`
+	ProxyServer string          `toml:"proxy_server"`
```
Comment (srebhan) [inline plugins/outputs/sematext/sematext.go:27]: Please use `plugins/common/proxy` here.

File: plugins/outputs/sematext/sematext.go [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+// SampleConfig Returns a sample configuration for the Sematext output
+func (s *Sematext) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns the description for the Sematext output
+func (s *Sematext) Description() string {
+	return "Use telegraf to send metrics to Sematext"
```
Comment (srebhan) [inline plugins/outputs/sematext/sematext.go:92]:
```suggestion
	return "Sends metrics to Sematext"
```

File: plugins/outputs/sematext/sematext.go [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+// Init performs full initialization of Sematext output
+func (s *Sematext) Init() error {
+	if len(s.Token) == 0 {
```
Comment (srebhan) [inline plugins/outputs/sematext/sematext.go:97]:
```suggestion
	if s.Token == "" {
```

File: plugins/outputs/sematext/sematext.go [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+func metricAlreadyProcessed(metric telegraf.Metric) bool {
+	_, processed := metric.GetTag(tags.SematextProcessedTag)
+	return processed
+}
+
+func markMetricsProcessed(metrics []telegraf.Metric) {
+	for _, m := range metrics {
+		m.AddTag(tags.SematextProcessedTag, tags.SematextProcessedTag)
+	}
+}
```
Comment (srebhan) [inline plugins/outputs/sematext/sematext.go:288]: I think using the metrics to track the sending state is completely flawed. Once you processed them they are dropped anyway, so why do you need to mark them?

File: plugins/outputs/sematext/sematext.go [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+		}
+		defer res.Body.Close()
+
+		s.Log.Debugf("Sending metrics to %s, response status code: %d", s.metricsURL, res.StatusCode)
+
+		return s.handleResponse(res)
+	}
+
+	return nil
+}
```
Comment (srebhan) [inline plugins/outputs/sematext/sematext.go:189]: I think you need a different approach here. Currently, you "misuse" the telegraf metrics to represent a Sematext specific structure by poking all that tags and fields. I'd rather suggest to define a Sematext specific metric type (e.g. `Datapoint` or whatever matches your terminology). Then iterate over the metrics and create and fill those `Datapoint`s using your processors. **But** those then should work on the defined Sematext structure and _not_ modify the telegraf metrics. Interpret it as a transformation to extract, collect, transform the metric values into something suited to be pushed to Sematext.

File: plugins/outputs/sematext/sender/sender.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
+package sender
```
Comment (srebhan) [inline plugins/outputs/sematext/sender/sender.go:1]: This whole sender package is Sematext specific and should be moved to an own library living outside of Telegraf.

File: plugins/outputs/sematext/serializer/escape.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
+// logic copied from plugins/serializers/influx/escape.go
+package serializer
```
Comment (srebhan) [inline plugins/outputs/sematext/serializer/escape.go:2]: Not Telegraf specific. Should move to an own library outside of Telegraf.

File: plugins/outputs/sematext/serializer/serializer.go [hunk 11]
```diff
@@ -0,0 +1,259 @@
+package serializer
```
Comment (srebhan) [inline plugins/outputs/sematext/serializer/serializer.go:1]: Not Telegraf specific. Should move to an own library outside of Telegraf.

**Review by srebhan — CHANGES_REQUESTED** [review]
Thank your contribution @bsmid and sorry for the late review! Regarding the code I have a few major comments:

1. There are quite some pieces here that are not Telegraf specific and should move to an own Sematext client library living outside of the Telegraf repository (e.g. `sender`). This will greatly reduce the amount of code in the PR.
2. Please change the way you process the metrics. Currently, you "transform" the Telegraf metrics "inplace" by poking tags and fields. I propose to define a Sematext specific structure that represents the way Sematext wants to see the data. Then map the metric values to the new structure using your processors. The difference would be that you can leave the Telegraf metrics unchanged and get a structure suited to be sent to Sematext. With the above, this structure could be defined in the external library.
3. Please avoid sending data in the Metainfo "processor". This is unexpected and makes the code more complicated.
4. Expecting fixed metric and field names is very fragile and flawed. Telegraf offers many possibilities to modify those (e.g. `name_override` or `processors`) and will thus break your assumption without any notice to the user. If this mapping (metric name to a "type") really necessary, provide a possibility for the user to _explicitly_ specify the mapping together with a sensible default.
4. Use the power of Go especially when dealing with `time` instead of implementing your own versions.

To summarize, I suggest to move all non-Telegraf specific code to an own, external library and then have another review round... How does that sound to you?

**Comment by bsmid** [comment]
Thanks, @srebhan, for very comprehensive feedback. I am not with Sematext anymore, though, so @otisg, could you please take this over?

**Comment by srebhan** [comment]
@otisg are you willing to work on this PR?

**Comment by otisg** [comment]
@srebhan we still have it on our TODO.... it's just that your comments look like the whole thing would have to be seriously redone.

**Comment by srebhan** [comment]
@otisg yeah I think the code requires some work. Good to know it's still on your list. Looking forward to the next round.

**Comment by srebhan** [comment]
Hey @otisg, given the need to rework this, I suggest to open a new PR once you are ready. Feel free to request a review or ping me. I'll close this one.
